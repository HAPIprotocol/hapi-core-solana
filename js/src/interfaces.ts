import {
  Commitment,
  Connection,
  ConnectionConfig,
  Keypair,
  PublicKey,
} from "@solana/web3.js";

export interface HapiClientConfig {
  endpoint: string | Connection;
  commitment?: Commitment | ConnectionConfig;
  communityName?: string;
  programId?: string;
}

export interface HapiViewResponse<T> {
  account: PublicKey;
  data: T;
}

export interface HapiActionResponse<T> {
  txHash: string;
  account: PublicKey;
  data: T;
}

export interface HapiActionResponseWithMeta<T, U> {
  txHash: string;
  account: PublicKey;
  data: T;
  meta: U;
}

export interface HapiClientAuthorityConfig extends HapiClientConfig {
  payer: Keypair | PublicKey;
}

export interface HapiClientReporterConfig extends HapiClientConfig {
  payer: Keypair | PublicKey;
}
