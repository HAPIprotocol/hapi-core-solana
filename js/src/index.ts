export * from "./state";
export * from "./constants";
export { ReaderClient, HapiClientConfig } from "./reader-client";
export { ReporterClient } from "./reporter-client";
export { AuthorityClient } from "./authority-client";
