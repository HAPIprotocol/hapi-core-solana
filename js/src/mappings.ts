import {
  HapiAccountType,
  HapiInstruction,
  ReporterType,
  Category,
  CaseStatus,
} from "./state/enums";

const HAPI_INSTRUCTION_NAMES = {
  [HapiInstruction.CreateCommunity]: "CreateCommunity",
  [HapiInstruction.UpdateCommunity]: "UpdateCommunity",
  [HapiInstruction.CreateNetwork]: "CreateNetwork",
  [HapiInstruction.UpdateNetwork]: "UpdateNetwork",
  [HapiInstruction.CreateReporter]: "CreateReporter",
  [HapiInstruction.UpdateReporter]: "UpdateReporter",
  [HapiInstruction.CreateCase]: "CreateCase",
  [HapiInstruction.UpdateCase]: "UpdateCase",
  [HapiInstruction.CreateAddress]: "CreateAddress",
  [HapiInstruction.UpdateAddress]: "UpdateAddress",
};
export function getHapiInstructionName(instruction: HapiInstruction): string {
  return HAPI_INSTRUCTION_NAMES[instruction] || "<Unknown>";
}

const HAPI_ACCOUNT_TYPE_NAMES = {
  [HapiAccountType.Uninitialized]: "Uninitialized",
  [HapiAccountType.Community]: "Community",
  [HapiAccountType.Network]: "Network",
  [HapiAccountType.Reporter]: "Reporter",
  [HapiAccountType.Case]: "Case",
  [HapiAccountType.Address]: "Address",
};
export function getHapiAccountTypeName(accountType: HapiAccountType): string {
  return HAPI_ACCOUNT_TYPE_NAMES[accountType] || "<Unknown>";
}

const REPORTER_TYPE_NAMES = {
  [ReporterType.Inactive]: "Inactive",
  [ReporterType.Tracer]: "Tracer",
  [ReporterType.Full]: "Full",
  [ReporterType.Authority]: "Authority",
};
export function getReporterTypeName(reporterType: ReporterType): string {
  return REPORTER_TYPE_NAMES[reporterType] || "<Unknown>";
}

const CATEGORY_NAMES = {
  [Category.Safe]: "Safe",
  [Category.WalletService]: "WalletService",
  [Category.MerchantService]: "MerchantService",
  [Category.MiningPool]: "MiningPool",
  [Category.LowRiskExchange]: "LowRiskExchange",
  [Category.MediumRiskExchange]: "MediumRiskExchange",
  [Category.DeFi]: "DeFi",
  [Category.OTCBroker]: "OTCBroker",
  [Category.ATM]: "ATM",
  [Category.Gambling]: "Gambling",
  [Category.IllicitOrganization]: "IllicitOrganization",
  [Category.Mixer]: "Mixer",
  [Category.DarknetService]: "DarknetService",
  [Category.Scam]: "Scam",
  [Category.Ransomware]: "Ransomware",
  [Category.Theft]: "Theft",
  [Category.Counterfeit]: "Counterfeit",
  [Category.TerroristFinancing]: "TerroristFinancing",
  [Category.Sanctions]: "Sanctions",
  [Category.ChildAbuse]: "ChildAbuse",
};
export function getCategoryName(category: Category): string {
  return CATEGORY_NAMES[category] || "<Unknown>";
}

const CASE_STATUS_NAMES = {
  [CaseStatus.Open]: "Open",
  [CaseStatus.Closed]: "Closed",
};
export function getCaseStatusName(caseStatus: CaseStatus): string {
  return CASE_STATUS_NAMES[caseStatus] || "<Unknown>";
}
