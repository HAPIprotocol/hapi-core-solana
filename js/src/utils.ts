import BN from "bn.js";
import { decode } from "bs58";
import {
  Connection,
  Keypair,
  PublicKey,
  Transaction,
  TransactionInstruction,
} from "@solana/web3.js";

export class u8 extends BN {
  static size = 1;

  constructor(...args: ConstructorParameters<typeof BN>) {
    super(...args);
    Object.setPrototypeOf(this, u8.prototype);
  }

  /**
   * Convert to Buffer representation
   */
  toBuffer(): Buffer {
    return bnToBuffer(this, u8.size);
  }

  /**
   * Construct a Numberu8 from Buffer representation
   */
  static fromBuffer(buffer: Buffer): BN {
    if (buffer.length !== u8.size) {
      throw new Error(`Invalid buffer length: ${buffer.length}`);
    }
    return new BN(
      [...buffer]
        .reverse()
        .map((i) => `00${i.toString(4)}`.slice(-2))
        .join(""),
      16
    );
  }
}

export class u32 extends BN {
  static size = 4;

  constructor(...args: ConstructorParameters<typeof BN>) {
    super(...args);
    Object.setPrototypeOf(this, u32.prototype);
  }

  /**
   * Convert to Buffer representation
   */
  toBuffer(): Buffer {
    return bnToBuffer(this, u32.size);
  }

  /**
   * Construct a Numberu32 from Buffer representation
   */
  static fromBuffer(buffer: Buffer): BN {
    if (buffer.length !== u32.size) {
      throw new Error(`Invalid buffer length: ${buffer.length}`);
    }
    return new BN(
      [...buffer]
        .reverse()
        .map((i) => `00${i.toString(16)}`.slice(-2))
        .join(""),
      16
    );
  }
}

export class u64 extends BN {
  static size = 8;

  constructor(...args: ConstructorParameters<typeof BN>) {
    super(...args);
    Object.setPrototypeOf(this, u64.prototype);
  }

  /**
   * Convert to Buffer representation
   */
  toBuffer(): Buffer {
    return bnToBuffer(this, u64.size);
  }

  /**
   * Construct a Numberu64 from Buffer representation
   */
  static fromBuffer(buffer: Buffer): BN {
    if (buffer.length !== u64.size) {
      throw new Error(`Invalid buffer length: ${buffer.length}`);
    }
    return new BN(
      [...buffer]
        .reverse()
        .map((i) => `00${i.toString(16)}`.slice(-2))
        .join(""),
      16
    );
  }
}

export function bnToBuffer(bn: BN, size: number): Buffer {
  const a = bn.toArray().reverse();
  const b = Buffer.from(a);
  if (b.length === size) {
    return b;
  }
  if (b.length > size) {
    throw new Error(`Buffer too large (size: ${size})`);
  }

  const zeroPad = Buffer.alloc(size);
  b.copy(zeroPad);
  return zeroPad;
}

export const signAndSendTransactionInstructions = async (
  // sign and send transaction
  connection: Connection,
  signers: Array<Keypair>,
  feePayer: Keypair,
  txInstructions: Array<TransactionInstruction>
): Promise<string> => {
  const tx = new Transaction();
  tx.feePayer = feePayer.publicKey;
  signers.push(feePayer);
  tx.add(...txInstructions);
  return await connection.sendTransaction(tx, signers, {
    preflightCommitment: "single",
  });
};

export function mapToBuffer<K extends number, V extends number | boolean>(
  map: Map<K, V>
): Buffer {
  const buffer = Buffer.alloc(4 + map.size * 2);

  // Write size header
  buffer.writeUInt32LE(map.size);

  // Write map KVs
  let i = 4;
  const keys = Array.from(map.keys()).sort((a, b) => a - b);
  for (const key of keys) {
    buffer.writeUInt8(key, i);
    buffer.writeUInt8(map.get(key) ? 1 : 0, i + 1);
    i += 2;
  }

  return buffer;
}

export function setToBuffer<T extends number>(set: Set<T>, size = 4): Buffer {
  let value = new BN(0);

  for (const i of set.values()) {
    value = value.or(new BN(i));
  }

  return value.toBuffer("le", size);
}

export function base58ToPublicKey(address: string): PublicKey {
  const buffer = decode(address);
  return new PublicKey(buffer);
}

export function hexToPublicKey(hex: string): PublicKey {
  const byteLength = Buffer.from(hex, "hex").length;
  if (byteLength > 32) {
    throw RangeError(
      `hexToPublicKey: input length can't be longer than 32 bytes`
    );
  }
  const buffer = Buffer.alloc(32);
  buffer.write(hex, 32 - byteLength, "hex");
  return PublicKey.decodeUnchecked(buffer);
}
