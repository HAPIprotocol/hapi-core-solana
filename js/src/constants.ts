import { PublicKey } from "@solana/web3.js";

export const HAPI_PROGRAM_ID = new PublicKey(
  "hapiScWyxeZy36fqXD5CcRUYFCUdid26jXaakAtcdZ7"
);
