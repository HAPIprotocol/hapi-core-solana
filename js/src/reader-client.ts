import {
  Commitment,
  Connection,
  ConnectionConfig,
  PublicKey,
} from "@solana/web3.js";

import { base58ToPublicKey, u64 } from "./utils";
import { Address, Case, Community, Network, Reporter } from "./state";
import { HAPI_PROGRAM_ID } from "./constants";

export interface HapiClientConfig {
  endpoint: string;
  commintment?: Commitment | ConnectionConfig;
  communityName?: string;
  programId?: string;
}

export interface HapiViewResponse<T> {
  account: PublicKey;
  data: T;
}

export interface HapiActionResponse<T> {
  account: PublicKey;
  txHash: string;
  data: T;
}

/** HAPI client to read entity data from Solana */
export class ReaderClient {
  protected connection: Connection;
  protected communityName?: string;
  protected programId: PublicKey;

  constructor(config: HapiClientConfig) {
    this.connection = new Connection(config.endpoint, config.commintment);
    this.communityName = config.communityName;
    this.programId = new PublicKey(this.programId || HAPI_PROGRAM_ID);
  }

  /**
   * Sets community name for context
   * @param communityName Community name
   * @returns Self
   */
  switchCommunity(communityName: string): ReaderClient {
    this.communityName = communityName;
    return this;
  }

  /**
   * Fetch community info from blockchain
   * @param communityName (Optional) The name of the community to fetch (defaults to context)
   * @returns Community info
   **/
  async getCommunity(
    communityName?: string
  ): Promise<HapiViewResponse<Community>> {
    if (!communityName && !this.communityName) {
      throw new Error("Community name not specified");
    }

    const state = await Community.retrieve(
      this.programId,
      this.connection,
      communityName || this.communityName
    );

    return state;
  }

  /**
   * Fetch network info from blockchain
   * @param networkName The name of the network to fetch
   * @param communityName (Optional) The name of the community to fetch network from (defaults to context)
   * @returns Network info
   **/
  async getNetwork(
    networkName: string,
    communityName?: string
  ): Promise<HapiViewResponse<Network>> {
    if (!communityName && !this.communityName) {
      throw new Error("Community name not specified");
    }

    if (!networkName) {
      throw new Error("Network name not specified");
    }

    const state = await Network.retrieve(
      this.programId,
      this.connection,
      communityName || this.communityName,
      networkName
    );

    return state;
  }

  /**
   * Fetch reporter info from blockchain
   * @param reporterPubkey Public key of the reporter to fetch
   * @param communityName (Optional) The name of the community to fetch reporter from (defaults to context)
   * @returns Reporter info
   **/
  async getReporter(
    reporterPubkey: string,
    communityName?: string
  ): Promise<HapiViewResponse<Reporter>> {
    if (!communityName && !this.communityName) {
      throw new Error("Community name not specified");
    }

    const state = await Reporter.retrieve(
      this.programId,
      this.connection,
      communityName || this.communityName,
      new PublicKey(reporterPubkey)
    );

    return state;
  }

  /**
   * Fetch case info from blockchain
   * @param caseId ID of the case to fetch
   * @param communityName (Optional) The name of the community to fetch case from (defaults to context)
   * @returns Case info
   **/
  async getCase(
    caseId: u64,
    communityName?: string
  ): Promise<HapiViewResponse<Case>> {
    if (!communityName && !this.communityName) {
      throw new Error("Community name not specified");
    }

    const state = await Case.retrieve(
      this.programId,
      this.connection,
      communityName || this.communityName,
      caseId
    );

    return state;
  }

  /**
   * Fetch address info from blockchain
   * @param address The address to fetch info for (string for Solana addresses, Buffer for others)
   * @param networkName The name of the network to which address belongs to
   * @param communityName (Optional) The name of the community to fetch case from (defaults to context)
   * @returns Address info
   **/
  async getAddress(
    address: string | Buffer,
    networkName: string,
    communityName?: string
  ): Promise<HapiViewResponse<Address>> {
    if (!communityName && !this.communityName) {
      throw new Error("Community name not specified");
    }

    if (!networkName) {
      throw new Error("Network name not specified");
    }

    const convertedAddress =
      address instanceof Buffer
        ? new PublicKey(address)
        : base58ToPublicKey(address);

    const state = await Address.retrieve(
      this.programId,
      this.connection,
      communityName || this.communityName,
      networkName,
      convertedAddress
    );

    return state;
  }
}
