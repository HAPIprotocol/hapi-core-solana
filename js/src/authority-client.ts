import { ReaderClient } from "./reader-client";

export class AuthorityClient extends ReaderClient {
  async createCommunity(): Promise<void> {
    // TODO: create an instruction
  }

  async createNetwork(): Promise<void> {
    // TODO: create an instruction
  }

  async addReporter(): Promise<void> {
    // TODO: create an instruction
  }

  async updateReporter(): Promise<void> {
    // TODO: create an instruction
  }
}
