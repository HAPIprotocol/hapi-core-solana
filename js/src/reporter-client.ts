import { PublicKey } from "@solana/web3.js";

import { HapiActionResponse, ReaderClient } from "./reader-client";
import { Case, Category } from "./state";

/** HAPI client to operate reporter program functions on Solana */
export class ReporterClient extends ReaderClient {
  async createCase(
    name: string,
    categories: Category[],
    communityName?: string
  ): Promise<HapiActionResponse<Case>> {
    name;
    categories;
    communityName;
    // TODO: create an instruction
    return {
      account: new PublicKey("0"),
    };
  }

  async updateCase(): Promise<void> {
    // TODO: create an instruction
  }

  async createAddress(): Promise<void> {
    // TODO: create an instruction
  }

  async updateAddress(): Promise<void> {
    // TODO: create an instruction
  }
}
