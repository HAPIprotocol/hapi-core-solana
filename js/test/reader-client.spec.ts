import b58 from "b58";

import { ReaderClient } from "../src";
import { u64 } from "../src/utils";

describe("ReaderClient", () => {
  let client: ReaderClient;

  beforeEach(() => {
    client = new ReaderClient({
      endpoint: "http://localhost:8899",
    });
  });

  it("should initialize", () => {
    expect(client).toBeDefined();
  });

  describe("viewCommunity", () => {
    it("should throw - not found", async () => {
      await expect(() =>
        client.viewCommunity("community404")
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should respond - success", async () => {
      const response = await client.viewCommunity("hapi.one");
      expect(response).toMatchSnapshot();
    });
  });

  describe("viewNetwork", () => {
    it("should throw - not found", async () => {
      await expect(() =>
        client.viewNetwork("unnetwork", "hapi.one")
      ).rejects.toThrowErrorMatchingSnapshot();

      await expect(() =>
        client.viewNetwork("testcoin", "community404")
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should respond - success", async () => {
      const response = await client.viewNetwork("testcoin", "hapi.one");
      expect(response).toMatchSnapshot();
    });
  });

  describe("viewReporter", () => {
    it("should throw - not found", async () => {
      await expect(() =>
        client.viewReporter(
          "49B9UtxhbVBKaToxThwQqshbmf2ZfnuJZX3XGjTsNLvb",
          "hapi.one"
        )
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should respond - success", async () => {
      // Alice
      {
        const response = await client.viewReporter(
          "DzMkTkH6ms7hEzyHisFnLLc2WDJfBb9TNNaPDQ7ADHhy",
          "hapi.one"
        );
        expect(response).toMatchSnapshot();
      }

      // Bob
      {
        const response = await client.viewReporter(
          "J5sUnZKuB1a9izNWDb4JEQzdB3J6mhe3sP6Ai6YCiKAZ",
          "hapi.one"
        );
        expect(response).toMatchSnapshot();
      }

      // Carol
      {
        const response = await client.viewReporter(
          "E8E298Dfe79V97ngYkcxXo1BftwduxjkGUWWRePM1Ac2",
          "hapi.one"
        );
        expect(response).toMatchSnapshot();
      }
    });
  });

  describe("viewCase", () => {
    it("should throw - not found", async () => {
      await expect(() =>
        client.viewCase(new u64(404), "hapi.one")
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should respond - success", async () => {
      const response = await client.viewCase(new u64(0), "hapi.one");
      expect(response).toMatchSnapshot();
    });
  });

  describe("viewAddress", () => {
    it("should throw - not found (string)", async () => {
      await expect(() =>
        client.viewAddress("4o4", "testcoin", "hapi.one")
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should throw - not found (buffer)", async () => {
      await expect(() =>
        client.viewAddress(
          Buffer.from("deadcode", "hex"),
          "testcoin",
          "hapi.one"
        )
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should throw - malformed", async () => {
      await expect(() =>
        client.viewAddress("BAD BASE58", "testcoin", "hapi.one")
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should throw - invalid community name", async () => {
      await expect(() =>
        client.viewAddress(
          "2Yy2iSPJv4iEMyNkUX7ydFoufSmyPLMc8P9owJopFRew",
          "testcoin",
          "community404"
        )
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should respond - success (string)", async () => {
      const response = await client.viewAddress(
        "2Yy2iSPJv4iEMyNkUX7ydFoufSmyPLMc8P9owJopFRew",
        "testcoin",
        "hapi.one"
      );
      expect(response).toMatchSnapshot();
    });

    it("should respond - success (buffer)", async () => {
      const buffer = b58.decode("2Yy2iSPJv4iEMyNkUX7ydFoufSmyPLMc8P9owJopFRew");
      const response = await client.viewAddress(buffer, "testcoin", "hapi.one");
      expect(response).toMatchSnapshot();
    });
  });
});
