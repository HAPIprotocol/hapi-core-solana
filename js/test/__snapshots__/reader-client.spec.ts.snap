// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReaderClient viewAddress should respond - success (buffer) 1`] = `
Object {
  "state": Address {
    "accountType": 5,
    "caseId": "01",
    "category": 16384,
    "risk": 5,
  },
}
`;

exports[`ReaderClient viewAddress should respond - success (string) 1`] = `
Object {
  "state": Address {
    "accountType": 5,
    "caseId": "01",
    "category": 16384,
    "risk": 5,
  },
}
`;

exports[`ReaderClient viewAddress should throw - invalid community name 1`] = `"Address not found: \\"2Yy2iSPJv4iEMyNkUX7ydFoufSmyPLMc8P9owJopFRew\\" in network \\"testcoin\\" (7GiNHuwBKg6URcBJXXXRPRtBVgn5q42AiowBEc8uzqBH) in community \\"community404\\" (mJLCqGmUWWDzYcDVHRAjaDJocoPDLzeimfD7KzwnxFh)"`;

exports[`ReaderClient viewAddress should throw - malformed 1`] = `"Non-base58 character"`;

exports[`ReaderClient viewAddress should throw - not found (buffer) 1`] = `"Address not found: \\"111111111111111111111111111111Hwr\\" in network \\"testcoin\\" (2viJmmn2pEfd6cogyqdDGWS9YkrVdnx87L994Qo3GwLx) in community \\"hapi.one\\" (DgBtqgnzYRsUZP3PhX5rCLfNycTQQ8cp7eMseosUQ4Ja)"`;

exports[`ReaderClient viewAddress should throw - not found (string) 1`] = `"Address not found: \\"1111111111111111111111111111114o4\\" in network \\"testcoin\\" (2viJmmn2pEfd6cogyqdDGWS9YkrVdnx87L994Qo3GwLx) in community \\"hapi.one\\" (DgBtqgnzYRsUZP3PhX5rCLfNycTQQ8cp7eMseosUQ4Ja)"`;

exports[`ReaderClient viewCase should respond - success 1`] = `
Object {
  "state": Case {
    "accountType": 4,
    "categories": Array [],
    "name": "case0",
    "reporterKey": PublicKey {
      "_bn": "c0fdcf81e4689a79e0cb64d88098f5342910652a340ddc0703dbdb10386dc678",
    },
  },
}
`;

exports[`ReaderClient viewCase should throw - not found 1`] = `"Invalid case account provided"`;

exports[`ReaderClient viewCommunity should respond - success 1`] = `
Object {
  "state": Community {
    "accountType": 1,
    "authority": PublicKey {
      "_bn": "ef37a50404b188a96e10f551122753f8ae8906c4ae0983c92727acf25b38f951",
    },
    "name": "hapi.one",
    "nextCaseId": "02",
  },
}
`;

exports[`ReaderClient viewCommunity should throw - not found 1`] = `"Community not found: community404 (mJLCqGmUWWDzYcDVHRAjaDJocoPDLzeimfD7KzwnxFh)"`;

exports[`ReaderClient viewNetwork should respond - success 1`] = `
Object {
  "state": Network {
    "accountType": 2,
    "name": "testcoin",
  },
}
`;

exports[`ReaderClient viewNetwork should throw - not found 1`] = `"Network not found: \\"unnetwork\\" (3TGzDJHkCHX5R62wtVh1Jr5TNe775jvSCt8tThbizBKQ) in community \\"hapi.one\\" (DgBtqgnzYRsUZP3PhX5rCLfNycTQQ8cp7eMseosUQ4Ja)"`;

exports[`ReaderClient viewNetwork should throw - not found 2`] = `"Network not found: \\"testcoin\\" (7GiNHuwBKg6URcBJXXXRPRtBVgn5q42AiowBEc8uzqBH) in community \\"community404\\" (mJLCqGmUWWDzYcDVHRAjaDJocoPDLzeimfD7KzwnxFh)"`;

exports[`ReaderClient viewReporter should respond - success 1`] = `
Object {
  "state": Reporter {
    "accountType": 3,
    "name": "Alice",
    "reporterType": 3,
  },
}
`;

exports[`ReaderClient viewReporter should respond - success 2`] = `
Object {
  "state": Reporter {
    "accountType": 3,
    "name": "Bob",
    "reporterType": 2,
  },
}
`;

exports[`ReaderClient viewReporter should respond - success 3`] = `
Object {
  "state": Reporter {
    "accountType": 3,
    "name": "Carol",
    "reporterType": 1,
  },
}
`;

exports[`ReaderClient viewReporter should throw - not found 1`] = `"Reporter not found: \\"49B9UtxhbVBKaToxThwQqshbmf2ZfnuJZX3XGjTsNLvb\\" in community \\"hapi.one\\" (DgBtqgnzYRsUZP3PhX5rCLfNycTQQ8cp7eMseosUQ4Ja)"`;
